From 9b8fa09823789c2281b193a9bce3cca5e70e6b84 Mon Sep 17 00:00:00 2001
From: Alain Volmat <alain.volmat@foss.st.com>
Date: Mon, 27 Mar 2023 10:23:17 +0200
Subject: [PATCH 496/529] media: i2c: gc2145: addition of all 4 Raw Bayer 8 bit
 formats

Addition of support for SBGGR8_1X8, SGBRG8_1X8, SRGGB8_1X8 and SGRBG8_1X8.

This correct as well the previously added support for SRGGB8_1X8 by
fixing the output format register field as described within the
specification document.

Signed-off-by: Alain Volmat <alain.volmat@foss.st.com>
Change-Id: I05c9cd78b3a5b5f08e2e36f0b3f9f7573ba307b7
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/linux-stm32/+/297504
Reviewed-by: Hugues FRUCHET <hugues.fruchet@st.com>
Reviewed-by: Philippe CORNU <philippe.cornu@foss.st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
Domain-Review: Philippe CORNU <philippe.cornu@foss.st.com>
---
 drivers/media/i2c/gc2145.c | 46 ++++++++++++++++++++++++++++++++++++--
 1 file changed, 44 insertions(+), 2 deletions(-)

--- a/drivers/media/i2c/gc2145.c
+++ b/drivers/media/i2c/gc2145.c
@@ -25,6 +25,9 @@
 /* Page 0 */
 #define GC2145_REG_ANALOG_MODE1	0x17
 #define GC2145_REG_OUTPUT_FMT	0x84
+#define GC2145_REG_SYNC_MODE	0x86
+#define GC2145_SYNC_MODE_COL_SWITCH	BIT(4)
+#define GC2145_SYNC_MODE_ROW_SWITCH	BIT(5)
 #define GC2145_REG_DEBUG_MODE2	0x8c
 #define GC2145_REG_DEBUG_MODE3	0x8d
 #define GC2145_REG_CHIP_ID	0xf0
@@ -894,12 +897,14 @@ static const struct gc2145_mode supporte
  * @colorspace: V4L2 colospace
  * @datatype: MIPI CSI2 data type
  * @output_fmt: GC2145 output format
+ * @row_col_switch: control of GC2145 row/col switch feature
  */
 struct gc2145_format {
 	unsigned int code;
 	unsigned int colorspace;
 	unsigned char datatype;
 	unsigned char output_fmt;
+	unsigned char row_col_switch;
 };
 
 /* All supported formats */
@@ -936,10 +941,32 @@ static const struct gc2145_format suppor
 		.output_fmt	= 0x06,
 	},
 	{
+		.code           = MEDIA_BUS_FMT_SGRBG8_1X8,
+		.colorspace     = V4L2_COLORSPACE_RAW,
+		.datatype       = MIPI_CSI2_DT_RAW8,
+		.output_fmt     = 0x17,
+		.row_col_switch = GC2145_SYNC_MODE_COL_SWITCH,
+	},
+	{
 		.code           = MEDIA_BUS_FMT_SRGGB8_1X8,
 		.colorspace     = V4L2_COLORSPACE_RAW,
 		.datatype       = MIPI_CSI2_DT_RAW8,
-		.output_fmt     = 0x19, /* Image is taken out of the Lens correction */
+		.output_fmt     = 0x17,
+		.row_col_switch = GC2145_SYNC_MODE_COL_SWITCH | GC2145_SYNC_MODE_ROW_SWITCH,
+	},
+	{
+		.code           = MEDIA_BUS_FMT_SBGGR8_1X8,
+		.colorspace     = V4L2_COLORSPACE_RAW,
+		.datatype       = MIPI_CSI2_DT_RAW8,
+		.output_fmt     = 0x17,
+		.row_col_switch = 0,
+	},
+	{
+		.code           = MEDIA_BUS_FMT_SGBRG8_1X8,
+		.colorspace     = V4L2_COLORSPACE_RAW,
+		.datatype       = MIPI_CSI2_DT_RAW8,
+		.output_fmt     = 0x17,
+		.row_col_switch = GC2145_SYNC_MODE_ROW_SWITCH,
 	},
 };
 
@@ -1281,6 +1308,7 @@ static int gc2145_start_streaming(struct
 	const struct gc2145_reg_list *reg_list;
 	const struct gc2145_format *gc2145_format;
 	uint16_t lwc, fifo_full_lvl, fifo_gate_mode;
+	u8 sync_mode;
 	int ret;
 
 	ret = pm_runtime_resume_and_get(&client->dev);
@@ -1313,6 +1341,17 @@ static int gc2145_start_streaming(struct
 	if (ret)
 		return ret;
 
+	ret = gc2145_read_reg(gc2145, GC2145_REG_SYNC_MODE, &sync_mode, 1);
+	if (ret)
+		return ret;
+
+	sync_mode &= ~(GC2145_SYNC_MODE_COL_SWITCH | GC2145_SYNC_MODE_ROW_SWITCH);
+	sync_mode |= gc2145_format->row_col_switch;
+
+	ret = gc2145_write_reg(gc2145, GC2145_REG_SYNC_MODE, sync_mode);
+	if (ret)
+		return ret;
+
 	/* Set 3rd page access */
 	ret = gc2145_write_reg(gc2145, GC2145_REG_PAGE_SELECT, 0x03);
 	if (ret)
@@ -1326,7 +1365,10 @@ static int gc2145_start_streaming(struct
 	 */
 	if (gc2145_format->colorspace != V4L2_COLORSPACE_RAW)
 		lwc = gc2145->mode->width * 2;
-	else if (gc2145_format->code == MEDIA_BUS_FMT_SRGGB8_1X8)
+	else if (gc2145_format->code == MEDIA_BUS_FMT_SGRBG8_1X8 ||
+		 gc2145_format->code == MEDIA_BUS_FMT_SRGGB8_1X8 ||
+		 gc2145_format->code == MEDIA_BUS_FMT_SBGGR8_1X8 ||
+		 gc2145_format->code == MEDIA_BUS_FMT_SGBRG8_1X8)
 		lwc = gc2145->mode->width * 1;
 	else
 		lwc = gc2145->mode->width + (gc2145->mode->width / 4);
